### PR TITLE
Task 4472, Allow rspconfig to set NTP servers for OpenBMC

### PR DESCRIPTION
**Pull request by xuweibj**
#4472 

Result of xcattest
```
------START::rspconfig_list_ntpservers::Time:Thu Dec 14 00:20:08 2017------

RUN:rspconfig f6u17 ntpservers [Thu Dec 14 00:20:08 2017]
ElapsedTime:1 sec
RETURN rc = 0
OUTPUT:
f6u17: BMC NTP Servers: None
CHECK:rc == 0	[Pass]
CHECK:output =~ f6u17: BMC NTP Servers	[Pass]

------END::rspconfig_list_ntpservers::Passed::Time:Thu Dec 14 00:20:09 2017 ::Duration::1 sec------
------Total: 1 , Failed: 0------

------START::rspconfig_set_ntpservers::Time:Thu Dec 14 00:48:50 2017------

RUN:/opt/xcat/share/xcat/tools/autotest/testcase/rspconfig/rspconfig_ntp.sh f6u17 c910f03c11k08 [Thu Dec 14 00:48:50 2017]
ElapsedTime:6 sec
RETURN rc = 0
OUTPUT:
The original BMC NTP Servers is None
Setting NTPServers as c910f03c11k08 success
To clear environment
Setting NTPServers as None success when clearing environment
CHECK:rc == 0	[Pass]

------END::rspconfig_set_ntpservers::Passed::Time:Thu Dec 14 00:48:56 2017 ::Duration::6 sec------
------Total: 1 , Failed: 0------
```

UT:
```
# rspconfig mid05tor12cn13,mid05tor12cn02 ntpservers=briggs01
mid05tor12cn13: BMC Setting NTPServers...
mid05tor12cn13: BMC NTP Servers: briggs01
mid05tor12cn02: BMC Setting NTPServers...
mid05tor12cn02: BMC NTP Servers: briggs01
# rspconfig mid05tor12cn13,mid05tor12cn02 ntpservers
mid05tor12cn13: BMC NTP Servers: briggs01
mid05tor12cn02: BMC NTP Servers: briggs01

# rspconfig mid05tor12cn13,mid05tor12cn05 ntpservers=
mid05tor12cn13: BMC Setting NTPServers...
mid05tor12cn05: BMC Setting NTPServers...
mid05tor12cn13: BMC NTP Servers: None
mid05tor12cn05: BMC NTP Servers: None
# rspconfig mid05tor12cn13,mid05tor12cn05 ntpservers
mid05tor12cn13: BMC NTP Servers: None
mid05tor12cn05: BMC NTP Servers: None
```

Run (get) with other options:
```
# rspconfig mid05tor12cn13 ntpservers ip
mid05tor12cn13: BMC NTP Servers: briggs01
mid05tor12cn13: BMC IP: 172.11.139.13
# rspconfig mid05tor12cn02 ntpservers ip
mid05tor12cn02: BMC NTP Servers: briggs01
mid05tor12cn02: Error: Interfaces with multiple IP addresses are not supported
# rspconfig mid05tor12cn13 ntpservers ip netmask gateway hostname
mid05tor12cn13: BMC NTP Servers: None
mid05tor12cn13: BMC IP: 172.11.139.13
mid05tor12cn13: BMC Netmask: 255.255.0.0
mid05tor12cn13: BMC Gateway: 0.0.0.0 (default: 172.11.253.27)
mid05tor12cn13: BMC Hostname: mid05tor12cn13-bmc-test
[root@briggs01 ~]# rspconfig mid05tor12cn02 ntpservers ip netmask gateway hostname
mid05tor12cn02: BMC NTP Servers: briggs01
mid05tor12cn02: BMC Hostname: mid05tor12cn02-bmc-test
mid05tor12cn02: Error: Interfaces with multiple IP addresses are not supported
```

### CODE REVIEW

File: xCAT-server/lib/xcat/plugins/openbmc.pm [hunk 1]
```diff
@@ -2752,79 +2777,103 @@ sub rspconfig_response {
...
                 if ($#nics > 1) {
                     $addon_info = " for $nic";
                 }
-                push @address, "BMC IP$addon_info: $nicinfo{$nic}{address}";
-                push @ipsrc, "BMC IP Source$addon_info: $nicinfo{$nic}{ipsrc}";
+
+                if ($nicinfo{$nic}{ntpservers}) {
+                    push @ntpservers, "BMC NTP Servers$addon_info: $nicinfo{$nic}{ntpservers}";
+                } else {
+                    push @ntpservers, "BMC NTP Servers$addon_info Not Set";
```
Comment (whowutwut) [inline xCAT-server/lib/xcat/plugins/openbmc.pm]: Maybe suggest change to: 

`BMC NTP Servers$addon_info: None`

So it looks like:   
```
BMC NTP Servers: None
```

File: xCAT-test/autotest/testcase/rspconfig/cases0 [hunk 2]
```diff
@@ -41,6 +41,21 @@ check:rc==0
 check:output=~__GETNODEATTR($$CN,hcp)__: \w+
 end
 
+start:rspconfig_list_ntpservers
+description: rspconfig list ntpservers info
+Attribute: $$CN-The operation object of rspconfig command
+cmd:rspconfig $$CN ntpservers
+check:rc==0
+check:output=~$$CN: BMC NTP Servers:|BMC NTP Servers Not Set
```
Comment (whowutwut) [inline xCAT-test/autotest/testcase/rspconfig/cases0]: Would also have to change this check here..  if change above.

File: xCAT-test/autotest/testcase/rspconfig/rspconfig_ntp.sh [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+echo "To clear environment"
+
+output=`rspconfig $cn ntpservers=$ntpservers`
+if [ $? -ne 0 ]; then
+    echo "rspconfig $cn ntpservers=$ntpservers failed when clearing environment"  
+    exit 1
+fi
+
+if [[ $output =~ "$cn: BMC NTP Servers: $ntpservers" ]] || [[ $output =~ "$cn: BMC NTP Servers Not Set" ]]; then
```
Comment (whowutwut) [inline xCAT-test/autotest/testcase/rspconfig/rspconfig_ntp.sh]: Same here, would need to change that "Not Set"

**Comment by whowutwut** [comment]
@xuweibj  This is great! I'd like to merge this in for 2.13.9 release so we can get some use of the function.  

For the comment above, I think we should either change to `: None` or at least `: Not Set` so we have a `:` to be consistent ...  

Feel free to merge if other reviewers have no comment.

File: xCAT-server/lib/xcat/plugins/openbmc.pm [hunk 4]
```diff
@@ -2752,79 +2777,103 @@ sub rspconfig_response {
...
+                    $multiple_error = "Interfaces with multiple IP addresses are not supported";
+                    next;
+                }
+
+                push @address, "BMC IP$addon_info: ${ $nicinfo{$nic}{address} }[0]";
+                push @ipsrc, "BMC IP Source$addon_info: ${ $nicinfo{$nic}{ipsrc} }[0]";
                 if ($nicinfo{$nic}{address}) {
-                    my $mask_shift = 32 - $nicinfo{$nic}{prefix};
+                    my $mask_shift = 32 - ${ $nicinfo{$nic}{prefix} }[0];
                     my $decimal_mask = (2 ** $nicinfo{$nic}{prefix} - 1) << $mask_shift;
```
Comment (robin2008) [inline xCAT-server/lib/xcat/plugins/openbmc.pm]: Here `$nicinfo{$nic}{prefix}` need to be replaced too.

**Comment by robin2008** [comment]
As the code changes have many impacts on old feature (To use array but not scalar),  we need full regression on existing feature.

**Comment by whowutwut** [comment]
@robin2008 Does @hu-weihua have a good size testing bucket for doing this? 
This code change should really only affect the `rspconfig` test cases...  

I would say, we should merge this in, do the regression and hold off on the 2.13.9 release until we are comfortable that this change did not regress the existing function. 

This is the importance of having not only FVT test bucket, but UT buckets from the developers...   FVT should not be doing UT...

File: xCAT-server/lib/xcat/plugins/openbmc.pm [hunk 5]
```diff
@@ -978,15 +990,15 @@ sub parse_args {
...
+                return ([ 1, "Unsupported command: $command $key" ]) unless ($key =~ /^ip$|^netmask$|^gateway$|^hostname$|^vlan$|^admin_passwd$|^ntpservers$/);
+                return ([ 1, "The option '$key' can not work with other options." ]) if ($key =~ /^hostname$|^admin_passwd$|^ntpservers$/ and $num_subcommand > 1);
                 if ($key eq "admin_passwd") {
                     my $comma_num = $value =~ tr/,/,/;
                     return ([ 1, "Invalid parameter for option $key: $value" ]) if ($comma_num != 1);
                 }
 
                 my $nodes_num = @$noderange;
-                return ([ 1, "Invalid parameter for option $key" ]) unless ($value);
+                return ([ 1, "Invalid parameter for option $key" ]) if (!$value and $key ne ("ntpservers"));
```
Comment (zet809) [inline xCAT-server/lib/xcat/plugins/openbmc.pm]: So, it means we can set '' to ntpservers?

File: xCAT-server/lib/xcat/plugins/openbmc.pm [hunk 6]
```diff
@@ -1353,6 +1365,19 @@ sub parse_command_status {
                 $status_info{RSPCONFIG_GET_RESPONSE}{argv} = "hostname";
                 return 0;
             }
+            if ($subcommand =~ /^ntpservers=(.*)/) {
+                $next_status{LOGIN_RESPONSE} = "RSPCONFIG_GET_NIC_REQUEST";
+                $next_status{RSPCONFIG_GET_NIC_REQUEST} = "RSPCONFIG_GET_NIC_RESPONSE";
+                $next_status{RSPCONFIG_GET_NIC_RESPONSE} = "RSPCONFIG_SET_NTPSERVERS_REQUEST";
+                $next_status{RSPCONFIG_SET_NTPSERVERS_REQUEST} = "RSPCONFIG_SET_RESPONSE";
```
Comment (zet809) [inline xCAT-server/lib/xcat/plugins/openbmc.pm]: Why there is a RSPCONFIG_SET_RESPONSE after SET NTPSERVER?

File: xCAT-server/lib/xcat/plugins/openbmc.pm [hunk 7]
```diff
@@ -2752,79 +2777,99 @@ sub rspconfig_response {
...
-                        # We have already processed an entry with adapter information.
-                        # This must be a second entry. Display an error. Currently only supporting
-                        # an adapter with a single IP address set.
-                        $error = "Interfaces with multiple IP addresses are not supported";
-                        $node_info{$node}{cur_status} = "";
-                        # Terminate loop on this error unless we are looking for hostname to display
-                        last unless ($grep_string =~ /(.*)hostname(.*)/);
+                    if ($content{Address} eq $node_info{$node}{bmcip} and $node_info{$node}{cur_status} eq "RSPCONFIG_GET_NIC_RESPONSE") {
+                        $status_info{RSPCONFIG_SET_NTPSERVERS_REQUEST}{init_url} =~ s/#NIC#/$nic/g;
+                        if ($next_status{ $node_info{$node}{cur_status} }) {
```
Comment (zet809) [inline xCAT-server/lib/xcat/plugins/openbmc.pm]: We can use `RSPCONFIG_GET_NIC_RESPONSE` here directly to replace $node_info{$node}{cur_status}

File: xCAT-server/lib/xcat/plugins/openbmc.pm [hunk 8]
```diff
@@ -2837,7 +2882,13 @@ sub rspconfig_response {
                     push @output, @vlan;
                 }
             }
+
             xCAT::SvrUtils::sendmsg("$_", $callback, $node) foreach (@output);
+            if ($multiple_error and $mul_out) {
+                xCAT::SvrUtils::sendmsg([1, "$multiple_error"], $callback, $node);
+                $wait_node_num--;
+                return;
+            }
```
Comment (zet809) [inline xCAT-server/lib/xcat/plugins/openbmc.pm]: Pls modify lines from 2904 to 2908 that using $nicinfo{$nic} address,prefix,gateway,vlan.

File: xCAT-server/lib/xcat/plugins/openbmc.pm (hunk 5 above)
Comment (xuweibj) [reply xCAT-server/lib/xcat/plugins/openbmc.pm]: yes

**Pull request merged**